### PR TITLE
Add initial batch of unit tests for the React API

### PR DESCRIPTION
**Pull request by rgossiaux**
These could be improved / made more complete but I think it's a start, and it's more testing than the old endpoints had anyway.

### CODE REVIEW

**Comment by rgossiaux** [comment]
I think the test failure is legitimate -- @asdfryan in #300 it looks like you didn't handle the case when `puzzle.chat_room` is None

**Review by erwa — COMMENTED** [review]
I haven't looked closely at the tests you added, but just a thought: do we want to add JavaScript tests that directly test the React/Redux code? I think that would add additional robustness on top of testing the Python code.

**Comment by rgossiaux** [comment]
Eventually, yes. I don't want to spend the time on them before mystery hunt though because I think there are other more critical things to do, but I'm hoping to get to it after this mystery hunt. Added #307 to track.

File: api/tests.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
+from .test_helpers import SmallboardTestCase
+from rest_framework import status
+from puzzles.models import Puzzle, PuzzleTag
+
+if "unittest.util" in __import__("sys").modules:
+    # Show full diff in self.assertEqual.
+    __import__("sys").modules["unittest.util"]._MAX_LENGTH = 999999999
```
Comment (erwa) [inline api/tests.py:7]: This seems a little strange to include here. I think you can just add `-vv` when running `./manage.py test` to get more info including the full diff. I think it'd be better to increase the verbosity when running tests so it applies to all tests rather than making the change just for these api tests.

**Review by erwa — COMMENTED** [review]
Thanks for adding tests.  Had a few comments.

**Comment by asdfryan** [comment]
> I think the test failure is legitimate -- @asdfryan in #300 it looks like you didn't handle the case when `puzzle.chat_room` is None

Oops -- will fix this.

File: hunts/src/api.js [hunk 2]
```diff
@@ -21,12 +21,12 @@ function handleErrors(response) {
 }
 
 function getPuzzles(huntId) {
-  const puzzlesApiUrl = `/api/v1/hunt/${huntId}/puzzles`;
+  const puzzlesApiUrl = `/api/v1/hunts/${huntId}/puzzles`;
```
[the remark this answers is not in the document]
Comment (rgossiaux) [reply hunts/src/api.js:24]: Yes. Actually I think the URL scheme is kind of bad, there's no need for the /hunts/:id/ prefix on most of the endpoints

File: api/tests.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+        # Empty string name
+        response = self.create_puzzle({"name": "", "url": TEST_URL, "is_meta": False})
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+        # Malformatted URL (TODO)
+        # response = self.create_puzzle(
+        #     {"name": "test name", "url": "bad_url", "is_meta": False}
+        # )
+        # self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
[the remark this answers is not in the document]
Comment (rgossiaux) [reply api/tests.py:45]: Yeah, that was the idea / why I included this. If I merge this PR first (likely) I'll update the other one to uncomment this.

File: api/tests.py (hunk 1 above)
Comment (rgossiaux) [reply api/tests.py:7]: Oh nice, I didn't know about -v. But I think that doesn't prevent the test output from being truncated:

```
======================================================================
FAIL: test_create_puzzle (api.tests.ApiTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/src/smallboard/api/tests.py", line 49, in test_create_puzzle
    self.assertEqual(
AssertionError: {'id'[137 chars]G', 'tags': [], 'guesses': [], 'metas': [], 'f[25 chars]alse} != {'id'[137 chars]G', 'guesses': [], 'metas': [], 'feeders': [],[13 chars]alse}
```

Since this runs during import I believe it applies to all tests as long as you're including `api/tests.py` but it wouldn't apply if you ran eg `test answers`, so I could move it somewhere else (maybe api/test_helpers.py?)

Comment (erwa) [reply api/tests.py:7]: What if you use two `v`s as in literally `-vv` for extra verbosity?

Comment (rgossiaux) [reply api/tests.py:7]:
```
usage: manage.py test [-h] [--noinput] [--failfast] [--testrunner TESTRUNNER] [-t TOP_LEVEL]
                      [-p PATTERN] [--keepdb] [-r] [--debug-mode] [-d] [--parallel [N]]
                      [--tag TAGS] [--exclude-tag EXCLUDE_TAGS] [--pdb] [-b]
                      [-k TEST_NAME_PATTERNS] [--version] [-v {0,1,2,3}]
                      [--settings SETTINGS] [--pythonpath PYTHONPATH] [--traceback]
                      [--no-color] [--force-color]
                      [test_label [test_label ...]]
manage.py test: error: argument -v/--verbosity: invalid int value: 'v'
```

The previous comment was with the most permissive verbosity level (3).

File: api/tests.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        )
+
+    def test_delete_puzzle(self):
+        self.create_puzzle({"name": "test name", "url": TEST_URL})
+        puzzle = Puzzle.objects.get()
+        response = self.delete_puzzle(puzzle.pk)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.data, {})
+        self.assertEqual(Puzzle.objects.count(), 0)
```
Comment (asdfryan) [inline api/tests.py:79]: Another edge case worth testing: Trying to delete a puzzle that has already been deleted

File: api/tests.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+    def test_edit_puzzle(self):
+        self.create_puzzle({"name": "test name", "url": TEST_URL})
+        puzzle = Puzzle.objects.get()
+
```
Comment (asdfryan) [inline api/tests.py:84]: Edge cases: 
- Editing more than 1 field at a time
- Editing a non-editable field
- New name matches a puzzle already in DB

File: api/tests.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        self.assertEqual(puzzle.status, Puzzle.SOLVED)
+        self.assertEqual(len(puzzle.correct_answers()), 1)
+
+        response = self.delete_answer(puzzle.pk, guesses[1].pk)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        puzzle.refresh_from_db()
+        self.assertEqual(puzzle.status, Puzzle.SOLVING)
+        self.assertEqual(len(puzzle.correct_answers()), 0)
+
+    def test_edit_answer(self):
```
Comment (asdfryan) [inline api/tests.py:160]: Edge cases:
- Editing an answer to another answer that already exists
- Special characters / emojis

File: hunts/src/api.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (rgossiaux) [reply hunts/src/api.js:24]: I'm gonna update this in another PR after this because otherwise I think the merge conflicts will get a little messy. But I think I want to move:

/hunts/:id/puzzles/:id/tags => /puzzles/:id/tags
/hunts/:id/puzzles/:id/tags/:id => /puzzles/:id/tags/:id
/hunts/:id/puzzles/:id/answers => /puzzles/:id/answers
/hunts/:id/puzzles/:id/answers/:id => /puzzles/:id/answers/:id

Hopefully that's a little more readable. Then I'll change this api.js file to use a common /api/v1/ prefix

**Review by asdfryan — APPROVED** [review]
Thanks! I commented some other edge cases worth testing. Feel free to merge this in first though. I can add those later.

File: api/tests.py (hunk 6 above)
Comment (rgossiaux) [reply api/tests.py:160]: We should definitely test these but I think the answers model tests do/should take care of these ones

Comment (rgossiaux) [reply api/tests.py:160]: (though from the API side I think ideally we should still test that the API returns reasonable responses in these cases)

File: api/tests.py (hunk 1 above)
Comment (erwa) [reply api/tests.py:7]: I see... I am surprised at unittest's behavior.
I wonder if you can move this unittest configuration setup to the root directory and have it always be picked up?
If not, I would prefer not changing the MAX_LENGTH at all, since I feel this sort of configuration should be done globally (and put in a global place if so) rather than done in a specific test, even if it applies globally. But if you feel strongly for including it, I'm fine with keeping it, too.

File: hunts/src/api.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (erwa) [reply hunts/src/api.js:24]: Ok, that's fine with me

File: api/tests.py (hunk 1 above)
Comment (rgossiaux) [reply api/tests.py:7]: I'll just delete it, probably simplest that way.

File: api/tests.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (erwa) [reply api/tests.py:41]: Oops, I linked to the wrong PR before, I meant #293
